Building with support for both OpenNI and OpenNI2 results in linking conflict

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV => 3.1
- Operating System / Platform => Debian Jessie 32/64 Bit (possible affects other platforms)
- Compiler => gcc / g++

##### Detailed description

Related to the closed issue #4918, conflict arises when trying to build with OpenNI and OpenNI 2 support at the same time with the function ``cvCreateFileCapture_OpenNI``

##### Steps to reproduce

1. Clone git master branch (32df09c27599e6327f87ca3dd05871a4e117737f)
2. Set CMake to build OpenNI and OpenNI 2 support: ``WITH_OPENNI=ON WITH_OPENNI2=ON``
3. Make

Next output is from a parallelized make with 4 cores. Tested also with no-parallelized make.
```
[ 43%] Linking CXX shared library ../../lib/libopencv_videoio.so
CMakeFiles/opencv_videoio.dir/src/cap_openni2.cpp.o: In function `cvCreateFileCapture_OpenNI(char const*)':
cap_openni2.cpp:(.text._Z26cvCreateFileCapture_OpenNIPKc+0x0): multiple definition of `cvCreateFileCapture_OpenNI(char const*)'
CMakeFiles/opencv_videoio.dir/src/cap_openni.cpp.o:cap_openni.cpp:(.text._Z26cvCreateFileCapture_OpenNIPKc+0x0): first defined here
collect2: error: ld returned 1 exit status
modules/videoio/CMakeFiles/opencv_videoio.dir/build.make:372: recipe for target 'lib/libopencv_videoio.so.3.1.0' failed
make[2]: *** [lib/libopencv_videoio.so.3.1.0] Error 1
CMakeFiles/Makefile2:3958: recipe for target 'modules/videoio/CMakeFiles/opencv_videoio.dir/all' failed
make[1]: *** [modules/videoio/CMakeFiles/opencv_videoio.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
   
```

--- a/modules/videoio/src/cap.cpp
+++ b/modules/videoio/src/cap.cpp
@@ -340,6 +340,12 @@ CV_IMPL CvCapture * cvCreateFileCaptureWithPreference (const char * filename, in
         if (apiPreference) break;
 #endif
 
+#ifdef HAVE_OPENNI2
+    case CV_CAP_OPENNI2:
+        TRY_OPEN(result, cvCreateFileCapture_OpenNI2 (filename))
+        if (apiPreference) break;
+#endif
+
     case CV_CAP_IMAGES:
         TRY_OPEN(result, cvCreateFileCapture_Images (filename))
     }

--- a/modules/videoio/src/cap_openni2.cpp
+++ b/modules/videoio/src/cap_openni2.cpp
@@ -1086,7 +1086,7 @@ CvCapture* cvCreateCameraCapture_OpenNI2( int index )
     return 0;
 }
 
-CvCapture* cvCreateFileCapture_OpenNI( const char* filename )
+CvCapture* cvCreateFileCapture_OpenNI2( const char* filename )
 {
     CvCapture_OpenNI2* capture = new CvCapture_OpenNI2( filename );
 

--- a/modules/videoio/src/precomp.hpp
+++ b/modules/videoio/src/precomp.hpp
@@ -124,6 +124,7 @@ CvVideoWriter* cvCreateVideoWriter_MSMF( const char* filename, int fourcc,
 CvCapture* cvCreateCameraCapture_OpenNI( int index );
 CvCapture* cvCreateCameraCapture_OpenNI2( int index );
 CvCapture* cvCreateFileCapture_OpenNI( const char* filename );
+CvCapture* cvCreateFileCapture_OpenNI2( const char* filename );
 CvCapture* cvCreateCameraCapture_Android( int index );
 CvCapture* cvCreateCameraCapture_XIMEA( int index );
 CvCapture* cvCreateCameraCapture_AVFoundation(int index);

